CRM/Eck/BAO/EckEntityType.php

```diff
@@ -183,4 +197,109 @@ function ($custom_group) use ($sub_type_value) {
     ]);
   }
 
+  /**
+   * Implements hook_civicrm_entityTypes().
+   *
+   * @see CRM_Utils_Hook::entityTypes()
+   * @link https://docs.civicrm.org/dev/en/latest/hooks/hook_civicrm_entityTypes
+   */
+  public static function onEntityTypes(&$entityTypes) {
+    $eck_entity_types = CRM_Core_DAO::executeQuery(
+      'SELECT * FROM `civicrm_eck_entity_type`;'
+    )->fetchAll('id');
+
+    foreach ($eck_entity_types as $entity_type) {
+      // "CRM_Eck_DAO_*" is a virtual class name, the corresponding class does
+      // not exist. "CRM_Eck_DAO_Entity" is therefore defined as the controller
+      // class.
+      $entityTypes['CRM_Eck_DAO_' . $entity_type['name']] = [
+        'name' => 'Eck_' . $entity_type['name'],
+        'class' => 'CRM_Eck_DAO_Entity',
+        'table' => _eck_get_table_name($entity_type['name']),
+      ];
+    }
+  }
+
+  /**
+   * Implements hook_civicrm_pre().
+   *
+   * @see CRM_Utils_Hook::pre()
+   * @link https://docs.civicrm.org/dev/en/latest/hooks/hook_civicrm_pre
+   */
+  public static function onPreEckEntityType(GenericHookEvent $event) {
+    $eckTypeName = $event->id ? CRM_Core_DAO::getFieldValue('CRM_Eck_DAO_EckEntityType', $event->id) : NULL;
+
+    switch ($event->action) {
+      case 'create':
+        // Replace special characters in the name
+        if (!empty($event->params['name'])) {
+          $event->params['name'] = CRM_Utils_String::munge($event->params['name'], '_', 52);
+        }
+        break;
+
+      case 'edit':
+        // Do not allow entity type to be renamed, as the table name depends on it
+        if (isset($event->params['name']) && $event->params['name'] !== $eckTypeName) {
+          throw new Exception('Renaming an EckEntityType is not allowed.');
```

Comment:
Maybe use `RuntimeException`. Don't know if it make sense here and if you have considered: But as `params` are passed as reference you could set the name back to the expected one instead of throwing an exception.